vnext/Chakra/ChakraJsiRuntime.cpp

```diff
@@ -0,0 +1,862 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+
+#include "ChakraJsiRuntime.h"
+#include "ChakraJsiRuntimeArgs.h"
+
+#include <MemoryTracker.h>
+#include <cxxreact/MessageQueueThread.h>
+
+#include <mutex>
+#include <sstream>
+#include "UnicodeConversion.h"
+
+#ifdef HAS_SCRIPTSTORE
+#include <jsi/ScriptStore.h>
+#endif
+
+using namespace facebook::react;
+
+namespace facebook { 
+namespace jsi {
+namespace chakraruntime {
+
+/*static*/ std::once_flag ChakraJsiRuntime::s_runtime_version_init_flag_;
+/*static*/ uint64_t ChakraJsiRuntime::s_runtimeVersion = 0;
+
+ChakraJsiRuntime::ChakraJsiRuntime(ChakraJsiRuntimeArgs&& args)  noexcept
+  : args_ { std::move(args) } {
+  
+  JsRuntimeAttributes runtimeAttributes = JsRuntimeAttributeNone;
+
+  if (!runtimeArgs().enableJITCompilation) {
+    runtimeAttributes = static_cast<JsRuntimeAttributes>(runtimeAttributes | JsRuntimeAttributeDisableNativeCodeGeneration | JsRuntimeAttributeDisableExecutablePageAllocation);
+  }
+
+  if (JsCreateRuntime(runtimeAttributes, nullptr, &runtime_) != JsNoError) {
+    std::terminate();
+  }
+
+  setupMemoryTracker();
+  setupNativePromiseContinuation();
+
+  // Create an execution context
+  JsCreateContext(runtime_, &ctx_);
+  JsAddRef(ctx_, nullptr);
+
+  // Note :: We currently assume that the runtime will be created and exclusively used in a single thread.
+  JsSetCurrentContext(ctx_);
+
+  std::call_once(s_runtime_version_init_flag_, initRuntimeVersion);
+}
+
+ChakraJsiRuntime::~ChakraJsiRuntime() noexcept {
+  JsSetCurrentContext(JS_INVALID_REFERENCE);
+  JsRelease(ctx_, nullptr);
+
+  JsDisposeRuntime(runtime_);
+}
+
+void ChakraJsiRuntime::setupMemoryTracker() noexcept{
+
+  if (runtimeArgs().memoryTracker) {
+    size_t initialMemoryUsage = 0;
+    JsGetRuntimeMemoryUsage(runtime_, &initialMemoryUsage);
+    runtimeArgs().memoryTracker->Initialize(initialMemoryUsage);
+
+    if (runtimeArgs().runtimeMemoryLimit > 0)
+      JsSetRuntimeMemoryLimit(runtime_, runtimeArgs().runtimeMemoryLimit);
+
+    JsSetRuntimeMemoryAllocationCallback(runtime_, runtimeArgs().memoryTracker.get(),
+      [](void* callbackState, JsMemoryEventType allocationEvent, size_t allocationSize) -> bool
+    {
+      auto memoryTrackerPtr = static_cast<MemoryTracker*>(callbackState);
+      switch (allocationEvent)
+      {
+      case JsMemoryAllocate:
+        memoryTrackerPtr->OnAllocation(allocationSize);
+        break;
+
+      case JsMemoryFree:
+        memoryTrackerPtr->OnDeallocation(allocationSize);
+        break;
+
+      case JsMemoryFailure:
+      default:
+        break;
+      }
+
+      return true;
+    }
+    );
+  }
+}
+      
+void ChakraJsiRuntime::evaluateJavaScript(
+  std::unique_ptr<const jsi::Buffer> buffer,
+  const std::string& sourceURL) {
+
+#ifdef HAS_SCRIPTSTORE

```

Comment:
>HAS_SCRIPTSTORE [](start = 7, length = 15)

This doesn't seem to be set anywhere, because its untested or not done?

Reply:
It can be removed now.. The script store headers were just checked into react-native .. This check was to ensure the code builds even without the react-native version bumped.